### PR TITLE
WiP: jobdata: selftest and compatibility fixes

### PR DESCRIPTION
**Pull request by ldoktor**
Hello guys,

this adds some compatibility fixes to be able to load jobs produced by older Avocado versions and adds a selftest which attempts to load data generated by various released Avocado versions and reports how successful it was. __It currently fails and my question is whether and for what cause we want to go deeper into history (without any fixes we only load Avocado-51.0 jobs).__

### CODE REVIEW

**Comment by ldoktor** [comment]
[Add selftests for legacy `jobdata`](https://trello.com/c/sTJoiTwI/1049-add-selftests-for-legacy-jobdata)

File: avocado/core/jobdata.py [hunk 1]
```diff
@@ -41,6 +43,20 @@
...
+    Look for a class including compatibility workarounds
+    """
+    if module == "avocado.plugins.yaml_to_mux":
+        mod = __import__("avocado_varianter_yaml_to_mux", fromlist=[module])
+        return getattr(mod, name)
+    else:
+        mod = __import__(module)
+        mod = sys.modules[module]
+        klass = getattr(mod, name)
+        return klass
```
Comment (apahim) [inline avocado/core/jobdata.py]: No need to have this `klass` variable, but that's minor. Wonder why did you remove the absolute-imports-only (`level=0`) restriction from the original `Unpickler.find_class()`. Any specific reason?

File: avocado/core/jobdata.py [hunk 2]
```diff
@@ -142,7 +158,9 @@ def retrieve_variants(resultsdir):
     # old pickle-based dump
     # TODO: Remove when 36lts is discontinued
     with open(recorded_mux, 'r') as mux_file:
-        return pickle.load(mux_file)
+        unpickler = pickle.Unpickler(mux_file)
+        unpickler.find_class = _find_class
+        return unpickler.load()
```
Comment (apahim) [inline avocado/core/jobdata.py]: good

File: avocado/core/mux.py [hunk 3]
```diff
@@ -181,6 +181,10 @@ def __iter__(self):
         """
         if self.root is None:
             return
+
+        # TODO: Remove when 36lts is disconinued
+        if not hasattr(self, "variant_ids"):
+            self.variant_ids = self._get_variant_ids()
```
Comment (apahim) [inline avocado/core/mux.py]: ok

File: avocado/core/jobdata.py [hunk 4]
```diff
@@ -146,6 +146,26 @@ def retrieve_variants(resultsdir):
...
+                        for mux_tree in variant:
+                            for node in mux_tree.pools:
+                                if hasattr(node, "filters"):
+                                    return  # it's new (or renewed) style
+                                root = node.root
+                                for _node in root.iter_children_preorder():
+                                    _node.filters = [[], []]
+                                    _node._environment = None
+        except:
+            pass
```
Comment (apahim) [inline avocado/core/jobdata.py]: Ok, hack-ish, but functional. I think we should clearly mark/comment those compatibility hacks so we can re-evaluate and possible remove them when we consider appropriate.

File: avocado/core/jobdata.py (hunk 1 above)
Comment (ldoktor) [inline avocado/core/jobdata.py]: Yep, it's a remnant from development. I'll return the getattr in v2. What do you mean by the `level=0` restriction? The original method on my system looks like this:

```python
    def find_class(self, module, name):
        # Subclasses may override this
        __import__(module)
        mod = sys.modules[module]
        klass = getattr(mod, name)
        return klass
```

File: avocado/core/jobdata.py (hunk 4 above)
Comment (ldoktor) [inline avocado/core/jobdata.py]: Sure, I can include the reason from commit message to code as well.

**Comment by ldoktor** [comment]
Dne 21.6.2017 v 16:32 Amador Pahim napsal(a):
> *@apahim* requested changes on this pull request.
> 
> @ldoktor <https://github.com/ldoktor> I like this compatibility package 
> of changes. It basically makes the replay compatible with everything 
> since last LTS. A couple of minor observations though.
> Thanks.
Well this is not actually true, that is why I sent this WiP to ask you 
whether I should proceed until we support 36lts, or whether it's not 
worth it. This version adds compatibility code to go up to 47.0. Going 
further would result in more and more code coming in.

Another solution would be to only focus on 36lts and add compatibility 
with that and skip the releases in between.

So what do you think:

1. Should we include those compatibility fixes to allow replaying 
47.0-master jobs?
2. Should I attempt to add compatibility fixes to allow replaying 
36.0-47.0 as well?
3. If not (2), should I at least add compatibility fixes for 36.0 to 
allow seamless switch from last LTS?

As the time is ticking the (2) seams near to impossible to achieve, (3) 
is on a tight schedule and it'll depend on how hard that task will turn...

Lukáš

File: avocado/core/jobdata.py (hunk 1 above)
Comment (apahim) [inline avocado/core/jobdata.py]: Yes, my bad. I was looking to the upstream pickle version (https://github.com/python/cpython/blob/master/Lib/pickle.py#L1381-L1392).

**Comment by apahim** [comment]
@ldoktor yes indeed. I don't think we need compatibility with every single version before 47.0. I' d not expend time with that. But I like the idea of having LTS compatibility, if feasible. That's basically the reason I was particularly happy about this PR. I mistaken considered that LTS compat support was in place.
